File: fugaku_in_practice/hands_on/sample_code/06_llio/02_sharedtmp/fortran/main.f90
! Copyright 2025 Research Organization for Information Science and Technology
!------------------------------------------------------------------------------
 
    program main

    use mpi

    implicit none

    integer, parameter :: input_no = 10
    character(len=132) argv
    character(len=132) input_file
    character(len=132) output_file
    logical :: file_exist

    integer :: array_size
    real(kind=8), allocatable :: A(:)
    integer :: ii
    integer :: n_loop
    integer :: i_skip

    !!! for MPI
    integer :: ierr
    integer :: nprocs, myrank
    integer :: ifh
    integer(kind=MPI_OFFSET_KIND) offset
    integer, parameter :: main_rank = 0

    !!! for elapse time measurement
    real(kind=8) :: t1, t2
    real(kind=8) :: total1, total2

    call MPI_Init(ierr)
    call MPI_Comm_size(MPI_COMM_WORLD, nprocs, ierr)
    call MPI_Comm_rank(MPI_COMM_WORLD, myrank, ierr)

    if (command_argument_count() >= 1) then
       call get_command_argument(1, argv)
       input_file = adjustl(argv)

       open (unit=input_no, file=input_file, status='old', action='read', iostat=ierr)

       if (ierr /= 0) then
          write(0,'("File Open Error: ", A)') input_file
          call MPI_Finalize(ierr)
          stop
       endif

       read (input_no, *) array_size
       read (input_no, *) n_loop
       read (input_no, *) i_skip
       read (input_no, '(A)') output_file

       close(unit=input_no)
    else
       array_size = 20
       n_loop = 100
       i_skip = 5
       output_file = 'output.dat'
    endif

600 format(1x,2(A,I6,2x),(A,I10,2x),(F8.3,1x,A,2x),2(A,I6,2x))

    if (myrank == main_rank) then
       write(6,600) 'NPROCS=',nprocs, &
                    'MYRANK=',myrank, &
                    'DATA_SIZE=',array_size, dble(array_size)*8.d0/(1024.d0*1024.d0),'MiB', &
                    'NLOOP=',n_loop, &
                    'ISKIP=',i_skip
    endif

    allocate(A(1:array_size), stat=ierr)
    if (ierr /= 0) then
       write(0,'("Allocation Error: stat = ",i2)') ierr
       call MPI_Abort(MPI_COMM_WORLD, 90, ierr)
       stop
    endif

    do ii = 1, array_size
       A(ii) = 1.d0 - 1.d0 / real(ii, kind=8)
    enddo

    call MPI_Barrier(MPI_COMM_WORLD, ierr)
    total1 = MPI_Wtime()

    do ii = 1, n_loop

       if (mod(ii, i_skip) == 0) then

          call MPI_Barrier(MPI_COMM_WORLD, ierr)
          t1 = MPI_Wtime()

          offset = (array_size * 8.d0) * myrank
          call MPI_File_open(MPI_COMM_WORLD, output_file, MPI_MODE_WRONLY+MPI_MODE_CREATE, MPI_INFO_NULL, ifh, ierr)
          call MPI_File_write_at(ifh, offset, A, array_size, MPI_REAL8, MPI_STATUS_IGNORE, ierr)
          call MPI_File_close(ifh, ierr)

          call MPI_Barrier(MPI_COMM_WORLD, ierr)
          t2 = MPI_Wtime()

          if (myrank == main_rank) &
             write(6,'(4x,"II=",i6,2x,"ELAPSED(SEC)=",1PE15.6)') ii, t2-t1
       endif
    enddo

    call MPI_Barrier(MPI_COMM_WORLD, ierr)
    total2 = MPI_Wtime()
    if (myrank == main_rank) &
       write(6,'("NPROCS=",i6,2x,"TOTAL ELAPSED(SEC)=",1PE15.6)') nprocs, total2 - total1

    deallocate(A)
    call MPI_Finalize(ierr)

end program main

File: fugaku_in_practice/hands_on/sample_code/06_llio/02_sharedtmp/cpp/main.cpp
/* Copyright 2025 Research Organization for Science and Technology */

#include <mpi.h>
#include <iostream>
#include <fstream>
#include <cstdlib>
#include <string>
#include <cmath>

int main(int argc, char* argv[])
{
   const int main_rank = 0;
   int myrank, nprocs;
   MPI_Offset offset;
   MPI_File ifh;

   int array_size;
   int i_skip, n_loop;
   std::string input_file, output_file;

   double *A;
   double t1, total1;
   double t2, total2;

   std::string tmp;

   MPI_Init(&argc, &argv);
   MPI_Comm_size(MPI_COMM_WORLD, &nprocs);
   MPI_Comm_rank(MPI_COMM_WORLD, &myrank);

   if (argc > 1) {
      input_file = argv[1];

      std::ifstream infil(input_file, std::ios::in);
      if (!infil){
         std::cerr << "File Open Error: " << input_file << std::endl;
         MPI_Finalize();
         return EXIT_FAILURE;
      }

      getline(infil, tmp);
      array_size = stoi(tmp);

      getline(infil, tmp);
      n_loop = stoi(tmp);

      getline(infil, tmp);
      i_skip = stoi(tmp);

      getline(infil, tmp);
      output_file = tmp;

      infil.close();
   } else {
      array_size = 20;
      n_loop = 100;
      i_skip = 5;
      output_file = "./output.txt";
   }

   if (myrank == main_rank) {
      std::cout << " NPROCS = " << nprocs << " ";
      std::cout << " MYRANK = " << myrank << " ";
      std::cout << " DATA_SIZE = " << array_size << " (" << array_size * 8.0 / (1024.0 * 1024.0) << " MiB) ";
      std::cout << " NLOOP = " << n_loop << " ";
      std::cout << " ISKIP = " << i_skip << std::endl;
   }

   try {
      A = new double[array_size];
   } catch (const std::bad_alloc& e){
      std::cerr << "Memory Allocation Error: " << e.what() << std::endl;
      MPI_Abort(MPI_COMM_WORLD, 90);
      return EXIT_FAILURE;
   }

   for (int ii = 0; ii < array_size; ii++) {
      A[ii] = 1.0 - 1.0 / (double)(ii+1);
   }

   MPI_Barrier(MPI_COMM_WORLD);
   total1 = MPI_Wtime();

   for (int ii = 0; ii < n_loop; ii++) {

      if (ii%i_skip == 0) {
         MPI_Barrier(MPI_COMM_WORLD);
         t1 = MPI_Wtime();

         offset = (array_size * 8.0) * myrank;
         MPI_File_open(MPI_COMM_WORLD, output_file.c_str(), MPI_MODE_WRONLY+MPI_MODE_CREATE, MPI_INFO_NULL, &ifh);
         MPI_File_write_at(ifh, offset, A, array_size, MPI_DOUBLE, MPI_STATUS_IGNORE);
         MPI_File_close(&ifh);

         MPI_Barrier(MPI_COMM_WORLD);
         t2 = MPI_Wtime();

         if (myrank == main_rank) {
            std::cout << "    II= " << ii << "  ELAPSED (SEC)= " << t2-t1 << std::endl; 
         }
      }
   }

   MPI_Barrier(MPI_COMM_WORLD);
   total2 = MPI_Wtime();
   if (myrank == main_rank) {
      std::cout << " NPROCS = " << nprocs << "  LOOP ELAPSED (SEC)= " << total2 - total1 << std::endl;
   }

   delete[] A;
   MPI_Finalize();
 
   return EXIT_SUCCESS;
}
